File: poker/hand_eval.py
"""Poker hand evaluation using UCI / Kaggle poker-hands-dataset encoding.

Dataset reference: https://www.kaggle.com/datasets/joogollucci/poker-hands-dataset
(UCI Machine Learning Repository — poker hand)

Encoding
--------
Suits (S1–S5): 1=Hearts, 2=Spades, 3=Diamonds, 4=Clubs
Ranks (C1–C5): 1=Ace … 10=Ten, 11=Jack, 12=Queen, 13=King
CLASS (0–9): hand category from high card through royal flush.
"""

from __future__ import annotations

from dataclasses import dataclass
from itertools import combinations
from typing import Iterable, Sequence

from blackjack.cards import Card, Rank, Suit

# UCI/Kaggle CLASS labels (0–9)
HAND_CLASS_NAMES: tuple[str, ...] = (
    "Nothing in hand",
    "One pair",
    "Two pairs",
    "Three of a kind",
    "Straight",
    "Flush",
    "Full house",
    "Four of a kind",
    "Straight flush",
    "Royal flush",
)

_UCI_SUIT_TO_SUIT: dict[int, Suit] = {
    1: Suit.HEARTS,
    2: Suit.SPADES,
    3: Suit.DIAMONDS,
    4: Suit.CLUBS,
}

_UCI_RANK_TO_RANK: dict[int, Rank] = {
    1: Rank.ACE,
    2: Rank.TWO,
    3: Rank.THREE,
    4: Rank.FOUR,
    5: Rank.FIVE,
    6: Rank.SIX,
    7: Rank.SEVEN,
    8: Rank.EIGHT,
    9: Rank.NINE,
    10: Rank.TEN,
    11: Rank.JACK,
    12: Rank.QUEEN,
    13: Rank.KING,
}

_RANK_TO_UCI: dict[Rank, int] = {v: k for k, v in _UCI_RANK_TO_RANK.items()}
_SUIT_TO_UCI: dict[Suit, int] = {v: k for k, v in _UCI_SUIT_TO_SUIT.items()}


@dataclass(frozen=True, slots=True)
class HandScore:
    """Comparable score for a five-card poker hand."""

    hand_class: int
    tiebreak: tuple[int, ...]

    @property
    def name(self) -> str:
        return HAND_CLASS_NAMES[self.hand_class]

    def as_tuple(self) -> tuple[int, ...]:
        return (self.hand_class, *self.tiebreak)


def card_to_uci(card: Card) -> tuple[int, int]:
    return (_SUIT_TO_UCI[card.suit], _RANK_TO_UCI[card.rank])


def card_from_uci(suit: int, rank: int) -> Card:
    return Card(_UCI_RANK_TO_RANK[rank], _UCI_SUIT_TO_SUIT[suit])


def cards_from_uci_row(row: Sequence[int]) -> list[Card]:
    """Parse a 10- or 11-field UCI dataset row into five ``Card`` objects."""
    if len(row) < 10:
        raise ValueError("UCI poker row needs at least 10 fields (S1,C1,…,S5,C5)")
    cards: list[Card] = []
    for i in range(0, 10, 2):
        cards.append(card_from_uci(int(row[i]), int(row[i + 1])))
    return cards


def evaluate_five_uci(suits: Sequence[int], ranks: Sequence[int]) -> HandScore:
    cards = [card_from_uci(s, r) for s, r in zip(suits, ranks, strict=True)]
    return evaluate_five_cards(cards)


def evaluate_five_cards(cards: Sequence[Card]) -> HandScore:
    if len(cards) != 5:
        raise ValueError("evaluate_five_cards expects exactly five cards")
    suits = [card_to_uci(c)[0] for c in cards]
    ranks = sorted((card_to_uci(c)[1] for c in cards), reverse=True)

    is_flush = len(set(suits)) == 1
    straight_high = _straight_high(ranks)
    is_straight = straight_high is not None

    counts: dict[int, int] = {}
    for r in ranks:
        counts[r] = counts.get(r, 0) + 1
    by_count = sorted(((cnt, r) for r, cnt in counts.items()), reverse=True)

    if is_straight and is_flush:
        if set(ranks) == {1, 10, 11, 12, 13}:
            return HandScore(9, (14,))
        return HandScore(8, (straight_high,))

    if by_count[0][0] == 4:
        quad_rank = by_count[0][1]
        kicker = max(r for r in ranks if r != quad_rank)
        return HandScore(7, (quad_rank, kicker))

    if by_count[0][0] == 3 and by_count[1][0] == 2:
        return HandScore(6, (by_count[0][1], by_count[1][1]))

    if is_flush:
        return HandScore(5, tuple(ranks))

    if is_straight:
        return HandScore(4, (straight_high,))

    if by_count[0][0] == 3:
        trips = by_count[0][1]
        kickers = sorted((r for r in ranks if r != trips), reverse=True)
        return HandScore(3, (trips, *kickers))

    if by_count[0][0] == 2 and by_count[1][0] == 2:
        high_pair, low_pair = sorted((by_count[0][1], by_count[1][1]), reverse=True)
        kicker = max(r for r in ranks if r not in (high_pair, low_pair))
        return HandScore(2, (high_pair, low_pair, kicker))

    if by_count[0][0] == 2:
        pair = by_count[0][1]
        kickers = sorted((r for r in ranks if r != pair), reverse=True)
        return HandScore(1, (pair, *kickers))

    return HandScore(0, tuple(ranks))


def _rank_values_for_straight(ranks: Sequence[int]) -> list[int]:
    """UCI ranks with Ace treated as 14 when forming broadway / high straights."""
    values = list(ranks)
    if 1 in values:
        values.append(14)
    return values


def _straight_high(ranks: Sequence[int]) -> int | None:
    expanded = _rank_values_for_straight(ranks)
    unique = sorted(set(expanded), reverse=True)
    if len(set(ranks)) != 5:
        return None
    # Broadway: A-K-Q-J-10 encoded as 1,13,12,11,10
    if set(ranks) == {1, 10, 11, 12, 13}:
        return 14
    # Wheel: A-2-3-4-5 encoded as 1,2,3,4,5
    if set(ranks) == {1, 2, 3, 4, 5}:
        return 5
    if len(unique) >= 5:
        top5 = sorted(set(ranks), reverse=True)
        if len(top5) == 5 and top5[0] - top5[4] == 4:
            return top5[0]
    return None


def best_hand_from_cards(cards: Sequence[Card]) -> tuple[HandScore, list[Card]]:
    """Return the best five-card score from up to seven hole + community cards."""
    if len(cards) < 5:
        raise ValueError("Need at least five cards")
    best_score: HandScore | None = None
    best_combo: list[Card] | None = None
    for combo in combinations(cards, 5):
        score = evaluate_five_cards(list(combo))
        if best_score is None or compare_scores(score, best_score) > 0:
            best_score = score
            best_combo = list(combo)
    assert best_score is not None and best_combo is not None
    return best_score, best_combo


def compare_scores(a: HandScore, b: HandScore) -> int:
    """Return positive if *a* beats *b*, negative if *b* beats *a*, else 0."""
    ta, tb = a.as_tuple(), b.as_tuple()
    if ta > tb:
        return 1
    if ta < tb:
        return -1
    return 0


def validate_against_dataset_class(cards: Sequence[Card], expected_class: int) -> bool:
    """Check that five cards match the dataset CLASS label (for tests)."""
    return evaluate_five_cards(list(cards)).hand_class == expected_class


# Sample rows from poker-hand-testing.data (UCI/Kaggle dataset)
DATASET_FIXTURES: tuple[tuple[tuple[int, ...], int], ...] = (
    ((1, 1, 1, 13, 2, 4, 2, 3, 1, 12), 0),
    ((3, 12, 3, 2, 3, 11, 4, 5, 2, 5), 1),
    ((4, 8, 1, 3, 2, 3, 2, 2, 2, 8), 2),
    ((1, 9, 1, 6, 4, 5, 3, 5, 1, 5), 3),
    ((1, 11, 1, 10, 1, 12, 3, 8, 1, 9), 4),
    ((1, 1, 1, 7, 1, 2, 1, 6, 1, 5), 5),
    ((4, 5, 1, 9, 3, 5, 4, 9, 2, 9), 6),
    ((1, 2, 3, 2, 4, 2, 2, 2, 1, 8), 7),
    ((4, 5, 4, 4, 4, 8, 4, 6, 4, 7), 8),
    ((4, 12, 4, 1, 4, 13, 4, 11, 4, 10), 9),
)

File: docs/js/holdem/hand_eval.js
/**
 * Poker hand evaluation — UCI / Kaggle poker-hands-dataset encoding.
 * @see https://www.kaggle.com/datasets/joogollucci/poker-hands-dataset
 */

export const HAND_CLASS_NAMES = [
  "Nothing in hand",
  "One pair",
  "Two pairs",
  "Three of a kind",
  "Straight",
  "Flush",
  "Full house",
  "Four of a kind",
  "Straight flush",
  "Royal flush",
];

const UCI_SUIT_TO_KEY = { 1: "H", 2: "S", 3: "D", 4: "C" };
const UCI_RANK_NAMES = ["", "A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"];

export function cardToUci(card) {
  const suitMap = { hearts: 1, spades: 2, diamonds: 3, clubs: 4 };
  const rankMap = { A: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, J: 11, Q: 12, K: 13 };
  return [suitMap[card.suit], rankMap[card.rank]];
}

export function cardsFromUciRow(row) {
  const cards = [];
  for (let i = 0; i < 10; i += 2) {
    cards.push({ suit: UCI_SUIT_TO_KEY[row[i]], rank: UCI_RANK_NAMES[row[i + 1]] });
  }
  return cards;
}

function straightHigh(ranks) {
  const set = new Set(ranks);
  if (set.size !== 5) return null;
  if (set.has(1) && set.has(10) && set.has(11) && set.has(12) && set.has(13)) return 14;
  if (set.has(1) && set.has(2) && set.has(3) && set.has(4) && set.has(5)) return 5;
  const sorted = [...set].sort((a, b) => b - a);
  if (sorted[0] - sorted[4] === 4) return sorted[0];
  return null;
}

export function evaluateFiveUci(suits, ranks) {
  const sorted = [...ranks].sort((a, b) => b - a);
  const isFlush = new Set(suits).size === 1;
  const sh = straightHigh(sorted);
  const isStraight = sh != null;

  const counts = {};
  for (const r of sorted) counts[r] = (counts[r] || 0) + 1;
  const byCount = Object.entries(counts).map(([r, c]) => [Number(c), Number(r)]).sort((a, b) => b[0] - a[0] || b[1] - a[1]);

  if (isStraight && isFlush) {
    if (new Set(ranks).has(1) && new Set(ranks).has(10) && new Set(ranks).has(11) && new Set(ranks).has(12) && new Set(ranks).has(13)) {
      return { handClass: 9, tiebreak: [14], name: HAND_CLASS_NAMES[9] };
    }
    return { handClass: 8, tiebreak: [sh], name: HAND_CLASS_NAMES[8] };
  }
  if (byCount[0][0] === 4) {
    const quad = byCount[0][1];
    const kicker = Math.max(...sorted.filter((r) => r !== quad));
    return { handClass: 7, tiebreak: [quad, kicker], name: HAND_CLASS_NAMES[7] };
  }
  if (byCount[0][0] === 3 && byCount[1][0] === 2) {
    return { handClass: 6, tiebreak: [byCount[0][1], byCount[1][1]], name: HAND_CLASS_NAMES[6] };
  }
  if (isFlush) return { handClass: 5, tiebreak: sorted, name: HAND_CLASS_NAMES[5] };
  if (isStraight) return { handClass: 4, tiebreak: [sh], name: HAND_CLASS_NAMES[4] };
  if (byCount[0][0] === 3) {
    const trips = byCount[0][1];
    const kickers = sorted.filter((r) => r !== trips).sort((a, b) => b - a);
    return { handClass: 3, tiebreak: [trips, ...kickers], name: HAND_CLASS_NAMES[3] };
  }
  if (byCount[0][0] === 2 && byCount[1][0] === 2) {
    const [hp, lp] = [byCount[0][1], byCount[1][1]].sort((a, b) => b - a);
    const kicker = Math.max(...sorted.filter((r) => r !== hp && r !== lp));
    return { handClass: 2, tiebreak: [hp, lp, kicker], name: HAND_CLASS_NAMES[2] };
  }
  if (byCount[0][0] === 2) {
    const pair = byCount[0][1];
    const kickers = sorted.filter((r) => r !== pair).sort((a, b) => b - a);
    return { handClass: 1, tiebreak: [pair, ...kickers], name: HAND_CLASS_NAMES[1] };
  }
  return { handClass: 0, tiebreak: sorted, name: HAND_CLASS_NAMES[0] };
}

export function evaluateFiveCards(cards) {
  const suits = [];
  const ranks = [];
  for (const c of cards) {
    const [s, r] = cardToUci(c);
    suits.push(s);
    ranks.push(r);
  }
  return evaluateFiveUci(suits, ranks);
}

function combos(arr, k) {
  if (k === 0) return [[]];
  if (arr.length < k) return [];
  const [first, ...rest] = arr;
  return [...combos(rest, k - 1).map((c) => [first, ...c]), ...combos(rest, k)];
}

export function bestHandFromCards(cards) {
  let best = null;
  let bestFive = null;
  for (const combo of combos(cards, 5)) {
    const score = evaluateFiveCards(combo);
    if (!best || compareScores(score, best) > 0) {
      best = score;
      bestFive = combo;
    }
  }
  return { score: best, cards: bestFive };
}

export function compareScores(a, b) {
  const ta = [a.handClass, ...a.tiebreak];
  const tb = [b.handClass, ...b.tiebreak];
  for (let i = 0; i < Math.max(ta.length, tb.length); i++) {
    if ((ta[i] ?? 0) > (tb[i] ?? 0)) return 1;
    if ((ta[i] ?? 0) < (tb[i] ?? 0)) return -1;
  }
  return 0;
}

export const DATASET_FIXTURES = [
  [[1, 1, 1, 13, 2, 4, 2, 3, 1, 12], 0],
  [[3, 12, 3, 2, 3, 11, 4, 5, 2, 5], 1],
  [[4, 8, 1, 3, 2, 3, 2, 2, 2, 8], 2],
  [[1, 9, 1, 6, 4, 5, 3, 5, 1, 5], 3],
  [[1, 11, 1, 10, 1, 12, 3, 8, 1, 9], 4],
  [[1, 1, 1, 7, 1, 2, 1, 6, 1, 5], 5],
  [[4, 5, 1, 9, 3, 5, 4, 9, 2, 9], 6],
  [[1, 2, 3, 2, 4, 2, 2, 2, 1, 8], 7],
  [[4, 5, 4, 4, 4, 8, 4, 6, 4, 7], 8],
  [[4, 12, 4, 1, 4, 13, 4, 11, 4, 10], 9],
];
